### ISSUE TITLE
Wrong MaxPool gradient?

### ISSUE THREAD

**Issue by yiyuezhuo** [FluxML/Flux.jl#1197]
Seems like MaxPool gives one gradient even for elements which contribute many results. For example, let `kernel-size=3, stride=2` and `padding=1`:

```
1  2  3  4  5  6
7  88 9  10 11 12
13 14 15 16 17 18
19 20 21 22 23 24
```

So 88 will contribute 4 elements on output while 12 will contribute just one. But it seems like that it's not the behavior for present code. A example compared to `PyTorch`:

```julia
using Flux
using PyCall

const torch = pyimport("torch")
const nn = pyimport("torch.nn")

maxpool_py = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
maxpool = MaxPool((3, 3); stride=2, pad=1)

dummy_input_py = torch.randn(2, 64, 112, 112)  # dummy ResNet feture map before MaxPool
dummy_input = permutedims(dummy_input_py.numpy(), [3,4,2,1]) |> copy
dummy_input_py.shape, dummy_input |> size, dummy_input_py.dtype, dummy_input |> typeof

block = maxpool
block_py = maxpool_py;

dummy_input_py.requires_grad_()

block_py(dummy_input_py).sum().backward()

ps = params(block)

gs = let block_let = block
    gradient(ps) do
        dummy_input |> block_let |> sum
    end
end

using Plots
theme(:dark)
```

```
gs.grads[first(keys(gs.grads))][:,:,1,1] |> heatmap
```
![image](https://user-images.githubusercontent.com/12798270/83282599-6b327e80-a20c-11ea-8651-5641d1c4d185.png)

```
get(get(dummy_input_py.grad, 0), 0).numpy() |> heatmap
```
![image](https://user-images.githubusercontent.com/12798270/83282574-5d7cf900-a20c-11ea-880f-ed8070f25fc2.png)

I know Flux's `BatchNorm` uses zero padding while PyTorch doesn't. So comparison will become more complex on boundary but the problem itself is still obvious to see.

My version info:

```
Julia Version 1.4.1
Commit 381693d3df* (2020-04-14 17:20 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, haswell)
```
```
  [587475ba] Flux v0.10.4
```

**Comment by yiyuezhuo** [comment]
It's actually a `NNlib` [problem](https://github.com/FluxML/NNlib.jl/blob/master/src/impl/pooling_direct.jl#L179), I found that removing `maxpool_already_chose` condition and replacing `                                    dx[x_idxs...] = dy_idx*alpha + beta*dx[x_idxs...]` with `dx[x_idxs...] +=dy_idx*alpha + beta*dx[x_idxs...]` could fix it. A PR is being developed.

The old implementation seems correct for basic `2x2` "non-overlap" `MaxPool`, as shown in Model Zoo vgg16 [example](https://github.com/FluxML/model-zoo/blob/master/vision/cifar10/cifar10.jl#L66), but it's not the case for listed example, which is used in ResNet.

**Comment by yiyuezhuo** [comment]
Fixed by 
https://github.com/FluxML/NNlib.jl/pull/207
